{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "96340cfa-9c6a-4388-85f6-4458d146a81f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting biopython\n",
      "  Downloading biopython-1.79-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n",
      "\u001b[K     |████████████████████████████████| 2.3 MB 36.0 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: numpy in /opt/conda/lib/python3.9/site-packages (from biopython) (1.20.3)\n",
      "Installing collected packages: biopython\n",
      "Successfully installed biopython-1.79\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9547040-4309-4845-827f-61b363bea522",
   "metadata": {},
   "source": [
    "### BWT ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19c50641-8e88-4824-8eff-fcdb3f798fe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Class used for performing Burrows-Wheeler transformation.'''\n",
    "from datetime import datetime\n",
    "\n",
    "def print_curr_datetime(message):\n",
    "    now = datetime.now()\n",
    "    dtString = now.strftime(\"%d/%m/%Y %H:%M:%S\")\n",
    "    print(str(message) + \" = \", dtString)\n",
    "\n",
    "class BWT:\n",
    "    transformedText = '' \n",
    "    saIndexes = []\n",
    "    tots = dict()\n",
    "    ranks = []\n",
    "    firstCol = dict()\n",
    "    \n",
    "    \n",
    "    def clear_parameters(self):\n",
    "        self.transformedText = '' \n",
    "        self.saIndexes = []\n",
    "        self.tots = dict()\n",
    "        self.ranks = []\n",
    "        self.firstCol = dict()\n",
    "        \n",
    "\n",
    "    def transform(self, text):\n",
    "        print_curr_datetime('BWT transform Begin')\n",
    "        '''Performs BW transformation and keeps the result in transformedText and saIndexes.\n",
    "        Also, initializes tots (pairs char:#), ranks (for transformedText).'''\n",
    "        self.clear_parameters()\n",
    "        self.transformedText = self.bwt_via_sa(text) # Initializes transformedText, saIndexes\n",
    "        self.rank_bwt()  # Initializes tots, ranks\n",
    "        self.calculate_first_col() # Initializes firstCol\n",
    "        print_curr_datetime('BWT transform End')\n",
    "        \n",
    "        \n",
    "    def suffix_array(self, t):\n",
    "        \"\"\" Given T return suffix array SA(T).  We use Python's sorted\n",
    "            function here for simplicity, but we can do better. \"\"\"\n",
    "        saMatrix = [(t[i:], i) for i in range(len(t))]\n",
    "        saMatrix.sort()\n",
    "        # Extract and return just the offsets\n",
    "        #     print(satups)\n",
    "        return list(map(lambda x: x[1], saMatrix))\n",
    "    \n",
    "    \n",
    "    def bwt_via_sa(self, t):\n",
    "        \"\"\" Given T, returns BWT(T) by way of the suffix array. \"\"\"\n",
    "        bw = []\n",
    "        self.saIndexes = self.suffix_array(t)\n",
    "        for si in self.saIndexes:\n",
    "            if si == 0: bw.append('$')\n",
    "            else: bw.append(t[si-1])\n",
    "        return ''.join(bw) # return string-ized version of list bw\n",
    "    \n",
    "    \n",
    "    def rank_bwt(self):\n",
    "        ''' Given BWT string bw, return parallel list of B-ranks.  Also\n",
    "        returns tots: map from character to # times it appears. \n",
    "        Initializes tots and ranks.'''\n",
    "        self.tots = dict()\n",
    "        self.ranks = []\n",
    "        for c in self.transformedText:\n",
    "            if c not in self.tots: self.tots[c] = 0\n",
    "            self.ranks.append(self.tots[c])\n",
    "            self.tots[c] += 1\n",
    "    \n",
    "    \n",
    "    def calculate_first_col(self):\n",
    "        ''' Return map from character to the range of rows prefixed by\n",
    "        the character. Initializes first column.'''\n",
    "        self.firstCol = {}\n",
    "        totc = 0\n",
    "        for c, count in sorted(self.tots.items()):\n",
    "            self.firstCol[c] = (totc, totc + count)\n",
    "            totc += count\n",
    "    \n",
    "    \n",
    "    def last_col_with_ranks(self):\n",
    "        ''' Returns list with tuples (rank, char) for lastColumn.'''\n",
    "        return list(zip(self.ranks, self.transformedText))\n",
    "    \n",
    "    \n",
    "    def original_text(self):\n",
    "        ''' Make T from BWT(T) '''\n",
    "        rowi = 0 # Start in first row\n",
    "        t = '$' # Start in rightmost character\n",
    "        while self.transformedText[rowi] != '$':\n",
    "            c = self.transformedText[rowi]\n",
    "            t = c + t # Prepend to answer\n",
    "            # Jump to row that starts with c of same ranke\n",
    "            rowi = self.firstCol[c][0] + self.ranks[rowi]\n",
    "        return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d9cf08ac-a58e-4b20-9dc6-87dba31a791d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BWT transform Begin =  30/05/2022 19:09:32\n",
      "BWT transform End =  30/05/2022 19:09:32\n",
      "Original: abaaba$\n",
      "\n",
      "First col\n",
      "{'$': (0, 1), 'a': (1, 5), 'b': (5, 7)}\n",
      "\n",
      "SaIndexes\n",
      "[6, 5, 2, 3, 0, 4, 1]\n",
      "\n",
      "Last col with ranks\n",
      "[(0, 'a'), (0, 'b'), (1, 'b'), (1, 'a'), (0, '$'), (2, 'a'), (3, 'a')]\n",
      "\n",
      "Tots.items()\n",
      "dict_items([('a', 4), ('b', 2), ('$', 1)])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# TESTS\n",
    "text = 'abaaba$'\n",
    "bwt = BWT()\n",
    "bwt.transform(text)\n",
    "original = bwt.original_text()\n",
    "\n",
    "print(\"Original: \" + original)\n",
    "print()\n",
    "# Expected: 'abaaba$'\n",
    "\n",
    "print('First col')\n",
    "print(bwt.firstCol)\n",
    "print()\n",
    "# Expected: {'$': (0, 1), 'a': (1, 5), 'b': (5, 7)}\n",
    "\n",
    "print('SaIndexes')\n",
    "print(bwt.saIndexes)\n",
    "print()\n",
    "# Expected: [6, 5, 2, 3, 0, 4, 1]\n",
    "\n",
    "print('Last col with ranks')\n",
    "print(bwt.last_col_with_ranks())\n",
    "print()\n",
    "# Expected: [(0, 'a'), (0, 'b'), (1, 'b'), (1, 'a'), (0, '$'), (2, 'a'), (3, 'a')]\n",
    "\n",
    "\n",
    "print('Tots.items()')\n",
    "print(bwt.tots.items())\n",
    "print()\n",
    "# Expected: dict_items([('a', 4), ('b', 2), ('$', 1)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd3ad25d-94c4-422e-8d22-06bb79f63db0",
   "metadata": {},
   "source": [
    "### FM Index ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e3960eed-1543-4b08-8ac4-1be77ff08e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FMIndex:\n",
    "    \n",
    "    bwt = {}\n",
    "    tally = dict() # Key = character, Value = occurences of character (len of BWT(T))\n",
    "    c = dict()  # Key = character, Value = (value for first row occurence)\n",
    "    \n",
    "    def __init__(self, text):\n",
    "        text = text + '$'\n",
    "        self.clear_parameters()\n",
    "        print('Perform BWT')\n",
    "        self.bwt = BWT() # We only care about L and SA\n",
    "        self.bwt.transform(text)\n",
    "        print('End BWT')\n",
    "        \n",
    "        print('Init. tally')\n",
    "        self.init_tally() # Initialize tally\n",
    "        print('Finish init. tally')\n",
    "        \n",
    "        print('Init. c')\n",
    "        self.init_c() # Initialize c\n",
    "        print('Finish init. c')\n",
    "        \n",
    "    \n",
    "    def clear_parameters(self):\n",
    "        self.bwt = {}\n",
    "        self.tally = dict()\n",
    "        self.c = dict()\n",
    "        \n",
    "    \n",
    "    def init_tally(self):\n",
    "        empty = []\n",
    "        occurences = dict()\n",
    "        for i in range(len(self.bwt.transformedText)):\n",
    "            empty.append(0)\n",
    "            occurences[self.bwt.transformedText[i]] = 0\n",
    "            i += 1\n",
    "            \n",
    "        for char in self.bwt.firstCol.keys():\n",
    "            self.tally[char] = empty.copy()\n",
    "            \n",
    "        for i in range(len(self.bwt.transformedText)):\n",
    "            character = self.bwt.transformedText[i]\n",
    "            occurences[character] += 1\n",
    "            \n",
    "            for char in self.tally.keys():\n",
    "                self.tally[char][i] = occurences[char]\n",
    "                \n",
    "\n",
    "    def init_c(self):\n",
    "        ''' Initializes c dictionary, where key = character \n",
    "        and value = tuple (index, value for first row occurence)'''\n",
    "        i = 0\n",
    "        for char in self.bwt.firstCol.keys():\n",
    "            self.c[char] = (i, self.bwt.firstCol[char][0])\n",
    "            i += 1\n",
    "            \n",
    "    \n",
    "    def query(self, pattern):\n",
    "        ''' Returns list of found positions for pattern in the text. \n",
    "        Rows are counted from 0 to n, not from 1.'''\n",
    "        positions = []\n",
    "        start = 1\n",
    "        end = len(self.bwt.transformedText) - 1\n",
    "        k = len(pattern) - 1\n",
    "        while k > -1:\n",
    "            ch = pattern[k]\n",
    "            # start = C[ch], end = C[ch + 1] with break if ch + 1 > all keys\n",
    "            if k == len(pattern) -1:\n",
    "                start = self.c[ch][1]\n",
    "                if self.c[ch][0] == len(self.c) -1:\n",
    "                    break\n",
    "                # Find C[ch + 1]\n",
    "                index = self.c[ch][0] + 1\n",
    "                newKey = ''\n",
    "                for key in self.c:\n",
    "                    if self.c[key][0] == index:\n",
    "                        newKey = key\n",
    "                        break\n",
    "                end = self.c[newKey][1] - 1\n",
    "            else:\n",
    "                start = self.c[ch][1] + self.tally[ch][start - 1]\n",
    "                end = self.c[ch][1] + self.tally[ch][end] - 1\n",
    "                \n",
    "            k -= 1\n",
    "        for i in range(start, end + 1):\n",
    "            positions.append(self.bwt.saIndexes[i])\n",
    "        positions.sort()\n",
    "        return positions.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "decfab6e-17e1-47b4-8b84-f586f8373d01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Perform BWT\n",
      "BWT transform Begin =  30/05/2022 19:09:39\n",
      "BWT transform End =  30/05/2022 19:09:39\n",
      "End BWT\n",
      "Init. tally\n",
      "Finish init. tally\n",
      "Init. c\n",
      "Finish init. c\n",
      "Tally:\n",
      "{'$': [0, 0, 0, 0, 1, 1, 1], 'A': [1, 1, 1, 1, 1, 2, 3], 'B': [0, 0, 0, 1, 1, 1, 1], 'N': [0, 1, 2, 2, 2, 2, 2]}\n",
      "\n",
      "C\n",
      "{'$': (0, 0), 'A': (1, 1), 'B': (2, 4), 'N': (3, 5)}\n",
      "\n",
      "Query for pattern: ANA\n",
      "[1, 3]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# TESTS\n",
    "# text = 'abaaba' # Example - Expected\n",
    "# text = 'GATGCGAGAGATG'\n",
    "text = 'BANANA'\n",
    "\n",
    "fmIndex = FMIndex(text)\n",
    "\n",
    "print('Tally:')\n",
    "print(fmIndex.tally)\n",
    "print()\n",
    "# Expected {'$': [0, 0, 0, 0, 1, 1, 1], 'a': [1, 1, 1, 2, 2, 3, 4], 'b': [0, 1, 2, 2, 2, 2, 2]}\n",
    "\n",
    "print('C')\n",
    "print(fmIndex.c)\n",
    "print()\n",
    "# Expected {'$': (0, 0), 'a': (1, 1), 'b': (2, 5)}\n",
    "\n",
    "# pattern = 'aba'\n",
    "# pattern = 'GAGA'\n",
    "pattern = 'ANA'\n",
    "print('Query for pattern: ' + pattern)\n",
    "print(fmIndex.query(pattern))\n",
    "print()\n",
    "\n",
    "# Expected\n",
    "# For text='GAGA' -> [5,7]\n",
    "# For text='ANA' -> [1,3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "643eb869-f33a-4994-a52d-df9842c167fa",
   "metadata": {},
   "source": [
    "### Global alignment ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fb3317a7-454e-4da2-9ccf-398441dbefa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "\n",
    "class Aligner:\n",
    "    match = 0\n",
    "    mismatch = 0\n",
    "    gap = 0\n",
    "    \n",
    "    def __init__(self, match, mismatch, gap):\n",
    "        ''' Match, mismatch, gap '''\n",
    "        self.match = match\n",
    "        self.mismatch = mismatch\n",
    "        self.gap = gap\n",
    "        \n",
    "\n",
    "    def scoring_matrix(self, a, b):\n",
    "        if a == b: return self.match\n",
    "        if a == '_' or b == '_' : return self.gap\n",
    "        return self.mismatch\n",
    "\n",
    "\n",
    "    def global_alignment(self, x, y):\n",
    "        ''' Accepts x, y '''\n",
    "        D = numpy.zeros((len(x) + 1, len(y) + 1), dtype=int)\n",
    "        \n",
    "        for i in range(1, len(x) + 1):\n",
    "            D[i,0] = D[i-1,0] + self.scoring_matrix(x[i-1], '_')  \n",
    "        for j in range(1, len(y)+1):\n",
    "            D[0,j] = D[0,j-1] + self.scoring_matrix('_', y[j-1])\n",
    "        \n",
    "        for i in range(1, len(x) + 1):\n",
    "            for j in range(1, len(y) + 1):\n",
    "                D[i,j] = max(D[i-1,j]   + self.scoring_matrix(x[i-1], '_'),\n",
    "                             D[i,j-1]   + self.scoring_matrix('_', y[j-1]), \n",
    "                             D[i-1,j-1] + self.scoring_matrix(x[i-1], y[j-1]))\n",
    "                \n",
    "        # function returns table and global alignment score\n",
    "        #alignment score is in cell (n,m) of the matrix\n",
    "        return D, D[len(x),len(y)] \n",
    "\n",
    "\n",
    "    def traceback(self, x, y, V):\n",
    "        ''' Accepts x, y, V '''\n",
    "        # initializing starting position cell(n,m)\n",
    "        i=len(x)\n",
    "        j=len(y)\n",
    "        \n",
    "        # initializing strings we use to represent alignments in x, y, edit transcript and global alignment\n",
    "        ax, ay, am, tr = '', '', '', ''\n",
    "        \n",
    "        # exit condition is when we reach cell (0,0)\n",
    "        while i > 0 or j > 0:\n",
    "            \n",
    "            # calculating diagonal, horizontal and vertical scores for current cell\n",
    "            d, v, h = -100, -100, -100\n",
    "            \n",
    "            if i > 0 and j > 0:\n",
    "                delta = 1 if x[i-1] == y[j-1] else 0\n",
    "                d = V[i-1,j-1] + self.scoring_matrix(x[i-1], y[j-1])  # diagonal movement   \n",
    "            if i > 0: v = V[i-1,j] + self.scoring_matrix(x[i-1], '_')  # vertical movement\n",
    "            if j > 0: h = V[i,j-1] + self.scoring_matrix('_', y[j-1])  # horizontal movement\n",
    "                \n",
    "            # backtracing to next (previous) cell\n",
    "            if d >= v and d >= h:\n",
    "                ax += x[i-1]\n",
    "                ay += y[j-1]\n",
    "                if delta == 1:\n",
    "                    tr += 'M'\n",
    "                    am += '|'\n",
    "                else:\n",
    "                    tr += 'R'\n",
    "                    am += ' '\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif v >= h:\n",
    "                ax += x[i-1]\n",
    "                ay += '_'\n",
    "                tr += 'D'\n",
    "                am += ' '\n",
    "                i -= 1\n",
    "            else:\n",
    "                ay += y[j-1]\n",
    "                ax += '_'\n",
    "                tr += 'I'\n",
    "                am += ' '\n",
    "                j -= 1\n",
    "                \n",
    "        alignment='\\n'.join([ax[::-1], am[::-1], ay[::-1]])\n",
    "        return alignment, tr[::-1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "781ce32a-9ae1-429f-a9b8-9dd1cacc5fea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACGTCA_GC\n",
      "|| |||| ||\n",
      "TATGTCATGC\n",
      "MMRMMMMIMM\n",
      "[[  0  -7 -14 -21 -28 -35 -42 -49 -56 -63 -70]\n",
      " [ -7   2  -5 -12 -19 -26 -33 -40 -47 -54 -61]\n",
      " [-14  -5   4  -3 -10 -17 -24 -31 -38 -45 -52]\n",
      " [-21 -12  -3   1  -6 -13 -15 -22 -29 -36 -43]\n",
      " [-28 -19 -10  -6   3  -4 -11 -18 -25 -27 -34]\n",
      " [-35 -26 -17  -8  -4   5  -2  -9 -16 -23 -30]\n",
      " [-42 -33 -24 -15 -11  -2   7   0  -7 -14 -21]\n",
      " [-49 -40 -31 -22 -18  -9   0   9   2  -5 -12]\n",
      " [-56 -47 -38 -29 -20 -16  -7   2   6   4  -3]\n",
      " [-63 -54 -45 -36 -27 -23 -14  -5  -1   3   6]]\n",
      "6\n",
      "MMRMMMMIMM\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n# Expected for match = 2, mismatch = -3, gap = -7: \\n\\nTACGTCA_GC\\n|| |||| ||\\nTATGTCATGC\\nMMRMMMMIMM\\n[[  0  -7 -14 -21 -28 -35 -42 -49 -56 -63 -70]\\n [ -7   2  -5 -12 -19 -26 -33 -40 -47 -54 -61]\\n [-14  -5   4  -3 -10 -17 -24 -31 -38 -45 -52]\\n [-21 -12  -3   1  -6 -13 -15 -22 -29 -36 -43]\\n [-28 -19 -10  -6   3  -4 -11 -18 -25 -27 -34]\\n [-35 -26 -17  -8  -4   5  -2  -9 -16 -23 -30]\\n [-42 -33 -24 -15 -11  -2   7   0  -7 -14 -21]\\n [-49 -40 -31 -22 -18  -9   0   9   2  -5 -12]\\n [-56 -47 -38 -29 -20 -16  -7   2   6   4  -3]\\n [-63 -54 -45 -36 -27 -23 -14  -5  -1   3   6]]\\n6\\nMMRMMMMIMM\\n'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TESTS\n",
    "\n",
    "x = 'TACGTCAGC'\n",
    "y = 'TATGTCATGC'\n",
    "\n",
    "\n",
    "match = 2 # 0 1 2\n",
    "mismatch = -3 # -3 -2\n",
    "gap = -7 # -5, -7\n",
    "\n",
    "aligner = Aligner(match, mismatch, gap)\n",
    "\n",
    "D, alignmentScore = aligner.global_alignment(x, y)\n",
    "alignment, transcript = aligner.traceback(x, y, D)\n",
    "\n",
    "print(alignment)\n",
    "print(transcript)\n",
    "print(D)\n",
    "print(alignmentScore)\n",
    "print(transcript)\n",
    "\n",
    "'''\n",
    "# Expected for match = 2, mismatch = -3, gap = -7: \n",
    "\n",
    "TACGTCA_GC\n",
    "|| |||| ||\n",
    "TATGTCATGC\n",
    "MMRMMMMIMM\n",
    "[[  0  -7 -14 -21 -28 -35 -42 -49 -56 -63 -70]\n",
    " [ -7   2  -5 -12 -19 -26 -33 -40 -47 -54 -61]\n",
    " [-14  -5   4  -3 -10 -17 -24 -31 -38 -45 -52]\n",
    " [-21 -12  -3   1  -6 -13 -15 -22 -29 -36 -43]\n",
    " [-28 -19 -10  -6   3  -4 -11 -18 -25 -27 -34]\n",
    " [-35 -26 -17  -8  -4   5  -2  -9 -16 -23 -30]\n",
    " [-42 -33 -24 -15 -11  -2   7   0  -7 -14 -21]\n",
    " [-49 -40 -31 -22 -18  -9   0   9   2  -5 -12]\n",
    " [-56 -47 -38 -29 -20 -16  -7   2   6   4  -3]\n",
    " [-63 -54 -45 -36 -27 -23 -14  -5  -1   3   6]]\n",
    "6\n",
    "MMRMMMMIMM\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cbf7a4c-1586-400f-8c3d-5c49b147dac9",
   "metadata": {},
   "source": [
    "### Seed&Extend ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8eb750ef-f398-4ebb-8029-d26f3e55022e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(read):\n",
    "    ''' read '''\n",
    "    reversedRead = []\n",
    "    i = len(read) - 1\n",
    "    while i >= 0:\n",
    "        if read[i] == 'A':\n",
    "            reversedRead.append('T')\n",
    "        elif read[i] == 'T':\n",
    "            reversedRead.append('A')\n",
    "        elif read[i] == 'C':\n",
    "            reversedRead.append('G')\n",
    "        elif read[i] == 'G':\n",
    "            reversedRead.append('C')\n",
    "        i -= 1\n",
    "    return reversedRead\n",
    "\n",
    "\n",
    "def seed_and_extend(referenceGenome, readId, isReverseComplement, read, seedLength, margin, aligner, fmIndex):\n",
    "    ''' referenceGenome, readId, read, seedLength, margin, aligner, fmIndex'''\n",
    "    results = []\n",
    "    seed = read[0:seedLength]\n",
    "    \n",
    "    seedPositions = fmIndex.query(seed)\n",
    "    \n",
    "    i = 0\n",
    "    for position in seedPositions:\n",
    "        # print(position)\n",
    "        start = position + seedLength\n",
    "        end = start + len(read) - seedLength + margin # end is not counted in slicing\n",
    "        if end >= len(referenceGenome):\n",
    "            continue\n",
    "            \n",
    "        alignedRef = referenceGenome[start:end]\n",
    "        alignedRead = read[seedLength:]\n",
    "        \n",
    "        D, alignmentScore = aligner.global_alignment(alignedRef, alignedRead)\n",
    "        alignment, transcript = aligner.traceback(alignedRef, alignedRead, D)\n",
    "        \n",
    "        # print('alignment completed for ' + str(i))\n",
    "        \n",
    "        i += 1\n",
    "        \n",
    "        results.append((readId, isReverseComplement, start - seedLength, alignmentScore, transcript))\n",
    "    \n",
    "    # print('seed_and_extend completed')\n",
    "    results.sort(key=lambda x: x[3], reverse=True)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "25b9423b-b366-4d0f-a215-47238d586be5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Perform BWT\n",
      "BWT transform Begin =  29/05/2022 15:54:30\n",
      "BWT transform End =  29/05/2022 15:54:30\n",
      "End BWT\n",
      "Init. tally\n",
      "Finish init. tally\n",
      "Init. c\n",
      "Finish init. c\n",
      "[('a', True, 4, -9, 'MMMMMDD'), ('a', True, 21, -9, 'MMMMDMD')]\n",
      "['G', 'C', 'T', 'T', 'G', 'C', 'T', 'C', 'C', 'C', 'T', 'G', 'A', 'C', 'T']\n"
     ]
    }
   ],
   "source": [
    "# TESTS\n",
    "\n",
    "referenceGenome = 'AAGAAGTCAGGGAGCAAGCAGAGTCAGGGAGCAAGCCACCAC'\n",
    "read = 'AGTCAGGGAGCAAGC'\n",
    "reversedRead = reverse_complement(read)\n",
    "seek_length = 10\n",
    "margin = 2\n",
    "aligner = Aligner(1, -3, -7)\n",
    "fmIndex = FMIndex(referenceGenome)\n",
    "results = seed_and_extend(referenceGenome, 'a', True, read, seek_length, margin, aligner, fmIndex)\n",
    "print(results)\n",
    "print(reversedRead)\n",
    "\n",
    "# Expected for referenceGenome = 'AAGAAGTCAGGGAGCAAGCAGAGTCAGGGAGCAAGCCACCAC'\n",
    "# read = 'AGTCAGGGAGCAAGC'\n",
    "# found seed positions: 4, 21\n",
    "# [('a', True, 4, -9, 'MMMMMDD'), ('a', True, 21, -9, 'MMMMDMD')]\n",
    "# reversed read \n",
    "# ['G', 'C', 'T', 'T', 'G', 'C', 'T', 'C', 'C', 'C', 'T', 'G', 'A', 'C', 'T']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ccff62f-7230-420d-b57e-c77bafeaa52f",
   "metadata": {},
   "source": [
    "### Main program ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "77f23207-71e5-4903-9fe5-7cc3707247c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read files\n",
      "Files read\n",
      "\n",
      "Begin fm index init.\n",
      "Perform BWT\n",
      "BWT transform Begin =  29/05/2022 16:37:28\n",
      "BWT transform End =  29/05/2022 16:37:28\n",
      "End BWT\n",
      "Init. tally\n",
      "Finish init. tally\n",
      "Init. c\n",
      "Finish init. c\n",
      "End fm index init.\n",
      "\n",
      "ANALYSIS START1-3-7 =  29/05/2022 16:37:28\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 16:41:42\n",
      "End datetime for read 20 =  29/05/2022 16:41:42\n",
      "Start datetime for reversed read 20 =  29/05/2022 16:41:42\n",
      "End datetime for reversed read 20 =  29/05/2022 16:41:42\n",
      "Start datetime for read 40 =  29/05/2022 16:45:51\n",
      "End datetime for read 40 =  29/05/2022 16:46:25\n",
      "Start datetime for reversed read 40 =  29/05/2022 16:46:25\n",
      "End datetime for reversed read 40 =  29/05/2022 16:46:25\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED1-3-7 =  29/05/2022 16:46:26\n",
      "ANALYSIS START1-3-5 =  29/05/2022 16:46:26\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 16:50:38\n",
      "End datetime for read 20 =  29/05/2022 16:50:38\n",
      "Start datetime for reversed read 20 =  29/05/2022 16:50:38\n",
      "End datetime for reversed read 20 =  29/05/2022 16:50:38\n",
      "Start datetime for read 40 =  29/05/2022 16:54:46\n",
      "End datetime for read 40 =  29/05/2022 16:55:20\n",
      "Start datetime for reversed read 40 =  29/05/2022 16:55:20\n",
      "End datetime for reversed read 40 =  29/05/2022 16:55:20\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED1-3-5 =  29/05/2022 16:55:20\n",
      "ANALYSIS START1-2-7 =  29/05/2022 16:55:20\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 16:59:34\n",
      "End datetime for read 20 =  29/05/2022 16:59:34\n",
      "Start datetime for reversed read 20 =  29/05/2022 16:59:34\n",
      "End datetime for reversed read 20 =  29/05/2022 16:59:34\n",
      "Start datetime for read 40 =  29/05/2022 17:03:42\n",
      "End datetime for read 40 =  29/05/2022 17:04:16\n",
      "Start datetime for reversed read 40 =  29/05/2022 17:04:16\n",
      "End datetime for reversed read 40 =  29/05/2022 17:04:16\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED1-2-7 =  29/05/2022 17:04:16\n",
      "ANALYSIS START1-2-5 =  29/05/2022 17:04:16\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 17:08:29\n",
      "End datetime for read 20 =  29/05/2022 17:08:29\n",
      "Start datetime for reversed read 20 =  29/05/2022 17:08:29\n",
      "End datetime for reversed read 20 =  29/05/2022 17:08:29\n",
      "Start datetime for read 40 =  29/05/2022 17:12:36\n",
      "End datetime for read 40 =  29/05/2022 17:13:10\n",
      "Start datetime for reversed read 40 =  29/05/2022 17:13:10\n",
      "End datetime for reversed read 40 =  29/05/2022 17:13:10\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED1-2-5 =  29/05/2022 17:13:10\n",
      "ANALYSIS START2-3-7 =  29/05/2022 17:13:10\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 17:17:24\n",
      "End datetime for read 20 =  29/05/2022 17:17:24\n",
      "Start datetime for reversed read 20 =  29/05/2022 17:17:24\n",
      "End datetime for reversed read 20 =  29/05/2022 17:17:24\n",
      "Start datetime for read 40 =  29/05/2022 17:21:32\n",
      "End datetime for read 40 =  29/05/2022 17:22:06\n",
      "Start datetime for reversed read 40 =  29/05/2022 17:22:06\n",
      "End datetime for reversed read 40 =  29/05/2022 17:22:06\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED2-3-7 =  29/05/2022 17:22:06\n",
      "ANALYSIS START2-3-5 =  29/05/2022 17:22:06\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 17:26:19\n",
      "End datetime for read 20 =  29/05/2022 17:26:19\n",
      "Start datetime for reversed read 20 =  29/05/2022 17:26:19\n",
      "End datetime for reversed read 20 =  29/05/2022 17:26:19\n",
      "Start datetime for read 40 =  29/05/2022 17:30:26\n",
      "End datetime for read 40 =  29/05/2022 17:31:00\n",
      "Start datetime for reversed read 40 =  29/05/2022 17:31:00\n",
      "End datetime for reversed read 40 =  29/05/2022 17:31:00\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED2-3-5 =  29/05/2022 17:31:00\n",
      "ANALYSIS START2-2-7 =  29/05/2022 17:31:00\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 17:35:13\n",
      "End datetime for read 20 =  29/05/2022 17:35:13\n",
      "Start datetime for reversed read 20 =  29/05/2022 17:35:13\n",
      "End datetime for reversed read 20 =  29/05/2022 17:35:13\n",
      "Start datetime for read 40 =  29/05/2022 17:39:21\n",
      "End datetime for read 40 =  29/05/2022 17:39:55\n",
      "Start datetime for reversed read 40 =  29/05/2022 17:39:55\n",
      "End datetime for reversed read 40 =  29/05/2022 17:39:55\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED2-2-7 =  29/05/2022 17:39:55\n",
      "ANALYSIS START2-2-5 =  29/05/2022 17:39:55\n",
      "Begin aligner init.\n",
      "End aligner init.\n",
      "\n",
      "Start reads\n",
      "Start datetime for read 20 =  29/05/2022 17:44:07\n",
      "End datetime for read 20 =  29/05/2022 17:44:08\n",
      "Start datetime for reversed read 20 =  29/05/2022 17:44:08\n",
      "End datetime for reversed read 20 =  29/05/2022 17:44:08\n",
      "Start datetime for read 40 =  29/05/2022 17:48:15\n",
      "End datetime for read 40 =  29/05/2022 17:48:49\n",
      "Start datetime for reversed read 40 =  29/05/2022 17:48:49\n",
      "End datetime for reversed read 40 =  29/05/2022 17:48:49\n",
      "Finish reads and all\n",
      "ANALYSIS FINISHED2-2-5 =  29/05/2022 17:48:49\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import sys\n",
    "import pandas as pd\n",
    "from datetime import datetime\n",
    "import os\n",
    "\n",
    "\n",
    "def readFASTA(fasta_file):\n",
    "  arr = []\n",
    "  for seq_record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    arr.append(seq_record.seq)\n",
    "  return arr[0]\n",
    "\n",
    "\n",
    "def readFASTQ(fastq_file):\n",
    "  arr = []\n",
    "  for seq_record in SeqIO.parse(fastq_file, \"fastq\"):\n",
    "    arr.append([seq_record.id, seq_record.seq])\n",
    "  return arr\n",
    "\n",
    "\n",
    "# fasta = 'reference'\n",
    "# fastq = 'reads'\n",
    "\n",
    "# print(readFASTA(fasta + '.fasta'))\n",
    "# print(readFASTQ(fastq + '.fastq')[0])\n",
    "\n",
    "\n",
    "def store_to_csv(data, exportFileName):\n",
    "    df = pd.DataFrame(data, columns =['read_id', 'is_rev_comp', 'position','alignment_score','transcript'])\n",
    "    df.to_csv(exportFileName)\n",
    "\n",
    "\n",
    "# Main program\n",
    "# def main():\n",
    "def main(fasta, fastq, match, mismatch, gap, seedlength, margin, fmIndex, referenceGenome, reads):\n",
    "    # To pass argv in Spyder -> Run>Configuration per file>Command line options. i.e: 2 3 a\n",
    "    # script, fasta, fastq, match, mismatch, gap, seedlength, margin = sys.argv\n",
    "    # To pass argv in Spyder -> Run>Configuration per file>Command line options. i.e: 2 3 a\n",
    "    # script, fasta, fastq, match, mismatch, gap, seedlength, margin = sys.argv\n",
    "    # match = int(match)\n",
    "    # mismatch= int(mismatch)\n",
    "    # gap = int(gap)\n",
    "    # seedlength = int(seedlength)\n",
    "    # margin = int(margin)\n",
    "    \n",
    "    # fasta = 'example_human_reference.fasta'\n",
    "    # fastq = 'medium_example_human_Illumina.pe_1.fastq'\n",
    "    # fasta = 'reference.fasta'\n",
    "    # fastq = 'reads.fastq'\n",
    "    # match = 2\n",
    "    # mismatch = -3\n",
    "    # gap = -7\n",
    "    # seedlength = 10\n",
    "    # margin = 2\n",
    "    \n",
    "    match = int(match)\n",
    "    mismatch= int(mismatch)\n",
    "    gap = int(gap)\n",
    "    seedlength = int(seedlength)\n",
    "    margin = int(margin)\n",
    "    \n",
    "    print_curr_datetime(\"ANALYSIS START\" + str(match) + str(mismatch) + str(gap))\n",
    "    \n",
    "    \n",
    "    # fasta += '.fasta'\n",
    "    # fastq += '.fastq'\n",
    "    # print('Read files')\n",
    "    # referenceGenome = readFASTA(fasta)\n",
    "    # reads = readFASTQ(fastq)\n",
    "    # print('Files read')\n",
    "    # print()\n",
    "\n",
    "        \n",
    "     \n",
    "    # print('Begin fm index init.')\n",
    "    # fmIndex = FMIndex(referenceGenome)\n",
    "    # print('End fm index init.')\n",
    "    # print()\n",
    "    \n",
    "    print('Begin aligner init.')\n",
    "    aligner = Aligner(match, mismatch, gap)\n",
    "    print('End aligner init.')\n",
    "    print()\n",
    "    \n",
    "    i = 1\n",
    "    print(\"Start reads\")\n",
    "    # fileName = 'results/result_read'\n",
    "\n",
    "    step = 20\n",
    "    bestResults = []\n",
    "    for read in reads:\n",
    "        if i % step == 0:\n",
    "            print_curr_datetime(\"Start datetime for read \" + str(i))\n",
    "        \n",
    "        results = seed_and_extend(referenceGenome, read[0], False, read[1], seedlength, margin, aligner, fmIndex)\n",
    "        if (len(results) > 0):\n",
    "            bestResults.append(results[0]) # For all\n",
    "            \n",
    "        # store_to_csv(results, fileName + str(i)+'.csv')\n",
    "        \n",
    "        if i % step == 0:\n",
    "            print_curr_datetime(\"End datetime for read \" + str(i))\n",
    "        \n",
    "        \n",
    "        if i % step == 0:\n",
    "            print_curr_datetime(\"Start datetime for reversed read \" + str(i))\n",
    "        \n",
    "        reverseCompRead = reverse_complement(read[1])\n",
    "        \n",
    "        results = seed_and_extend(referenceGenome, read[0], True, reverseCompRead, seedlength, margin, aligner, fmIndex)\n",
    "        if (len(results) > 0):\n",
    "            bestResults.append(results[0]) # For all\n",
    "            \n",
    "        # store_to_csv(results, fileName + 'reversed' + str(i)+'.csv')\n",
    "        \n",
    "        if i % step == 0:\n",
    "            print_curr_datetime(\"End datetime for reversed read \" + str(i))\n",
    "        \n",
    "        i += 1\n",
    "    \n",
    "    \n",
    "    store_to_csv(bestResults, outputdir + '/our_results' + str(match) + str(mismatch) + str(gap) + '.csv')\n",
    "    \n",
    "    print(\"Finish reads and all\")\n",
    "    \n",
    "    print_curr_datetime(\"ANALYSIS FINISHED\" + str(match) + str(mismatch) + str(gap))\n",
    "    \n",
    "    '''\n",
    "    # TESTS\n",
    "    referenceGenome = 'AAGAAGTCAGGGAGCAAGCAGAGTCAGGGAGCAAGCCACCAC'\n",
    "    read = 'AGTCAGGGAGCAAGC'\n",
    "    reversedRead = reverse_complement(read)\n",
    "    seek_length = 10\n",
    "    margin = 2\n",
    "\n",
    "    \n",
    "    \n",
    "    aligner = Aligner(1, -3, -7)\n",
    "    fmIndex = FMIndex(referenceGenome)\n",
    "    results = seed_and_extend(referenceGenome, read, seek_length, margin, aligner, fmIndex)\n",
    "    print(results)\n",
    "    '''\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    fasta = 'example_human_reference.fasta'\n",
    "    # fastq = 'medium_example_human_Illumina.pe_1.fastq'\n",
    "    # fasta = \"dummy.fasta\"\n",
    "    fastq = \"medium_example_human_Illumina.pe_1.fastq\"\n",
    "    seed = 10\n",
    "    margin = 2\n",
    "    \n",
    "    outputdir = 'results_ours'\n",
    "    if not os.path.exists(outputdir):\n",
    "        os.mkdir(outputdir)\n",
    "        \n",
    "    print('Read files')\n",
    "    referenceGenome = readFASTA(fasta)\n",
    "    reads = readFASTQ(fastq)\n",
    "    print('Files read')\n",
    "    print()\n",
    "    \n",
    "    print('Begin fm index init.')\n",
    "    fmIndex = FMIndex(referenceGenome)\n",
    "    print('End fm index init.')\n",
    "    print()\n",
    "    \n",
    "    for match in range(1,3):\n",
    "        for mismatch in range(-3,-1):\n",
    "            for gap in range(-7, -4):\n",
    "                if gap == -6:\n",
    "                    continue\n",
    "                main(fasta, fastq, match, mismatch, gap, seed, margin, fmIndex, referenceGenome, reads)\n",
    "    # main()\n",
    "\n",
    "    '''\n",
    "    for match in range(0,3):\n",
    "        for mismatch in range(-3,-1):\n",
    "            for gap in range(-7, -4):\n",
    "                if gap == -5:\n",
    "                    continue\n",
    "                aligner = Aligner(match, mismatch, gap)\n",
    "    '''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0f68d220-05ab-4cec-88f3-7892697f06c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "example_human_Illumina.pe_1.fastq  example_human_reference.fasta\n",
      "example_human_Illumina.pe_2.fastq\n"
     ]
    }
   ],
   "source": [
    "!ls /sbgenomics/project-files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ab1269a-c9ad-4bb4-9243-706629ed9f0a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
